stdlib/SparseArrays/src/sparsematrix.jl

```diff
@@ -3516,12 +3521,20 @@ function spdiagm_internal(kv::Pair{<:Integer,<:AbstractVector}...)
             col = 0
         end
         r = 1+i:numel+i
-        I[r] = row+1:row+numel
-        J[r] = col+1:col+numel
-        copyto!(view(V, r), vect)
+        if vect isa AbstractSparseVector
```

Comment:
Maybe it would be possible to translate this part into two modifying methods, e.g.,
```julia
_copytodiag!((I, J, V), vect::SparseVector, row, col)
_copytodiag!((I, J, V), vect::AbstractVector, row, col)
```
 as well?

Reply:
One can do that. One issue is that `AbstractSparseVector` is not available at this point, so the method needs to be defined in a different file, same as the `_numel` method for the sparse vectors. It seems to me that it might be hard to understand what this method is good for, because it's not called within `sparsevector.jl`. So I'm not sure, but wouldn't mind either.

Reply:
Hm, one also needs to pass `numinds` and `numel`/`r`. I actually considered that earlier, but found that one would need to pass what feels like dozens of variables. 😄

Reply:
Yeah, true. I did not realize that. Well, we could do something like
```julia
_numel(vect::AbstractSparseVector) = nnz(vect)
_numel(vect::AbstractVector) = length(vect)

function _elem_ix(vect::AbstractSparseVector, row, col)
    ix = nonzeroinds(vect)
    return (row .+ ix, col .+ ix)
end
function _elem_ix(vect::AbstractVector, row, col)
    numinds = length(vect)
    return (row+1:row+numinds, col+1:col+numinds)
end

_elems(vect::AbstractSparseVector) = nonzeros(vect)
_elems(vect::AbstractVector) = vect

function spdiagm_internal(kv::Pair{<:Integer,<:AbstractVector}...)
    ncoeffs = 0
    for p in kv
        ncoeffs += _numel(p.second)
    end
    I = Vector{Int}(undef, ncoeffs)
    J = Vector{Int}(undef, ncoeffs)
    V = Vector{promote_type(map(x -> eltype(x.second), kv)...)}(undef, ncoeffs)
    i = 0
    m = 0
    n = 0
    for p in kv
        dia = p.first
        vect = p.second
        numel = _numel(vect)
        if dia < 0
            row = -dia
            col = 0
        elseif dia > 0
            row = 0
            col = dia
        else
            row = 0
            col = 0
        end
        r = 1+i:numel+i
        I[r], J[r] = _elem_ix(vect, row, col)
        copyto!(view(V, r), _elems(vect))

        numinds = length(vect)
        m = max(m, row + numinds)
        n = max(n, col + numinds)
        i += numel
    end
    return I, J, V, m, n
end
```
But this approach introduces two additional "helper" functions so I am not sure if that is a good design...

Reply:
Basically, I went with your suggestion of using more helper functions. Thank you!